### PR TITLE
#5337: make Mixtral multidevice+async

### PR DESCRIPTION
**Pull request by sraizada-tt**
replacing device loops with ttnn multi device + async API

### CODE REVIEW

**Comment by mtairum** [comment]
Added a commit for some leftover code on my previous merge (oops).

Started all 3 multichip pipelines. We'll have to check that they are all passing the mixtral tests at least.

- https://github.com/tenstorrent/tt-metal/actions/runs/9098906802
- https://github.com/tenstorrent/tt-metal/actions/runs/9098912081
- https://github.com/tenstorrent/tt-metal/actions/runs/9098913480

File: models/demos/t3000/mixtral8x7b/tests/test_mixtral_model.py [hunk 1]
```diff
@@ -32,28 +34,25 @@ def forward(self, x):
...
+# )
+# @pytest.mark.parametrize(
+#     "n_layers",
+#     (1, 32),
+# )
+# @pytest.mark.parametrize(
+#     "iterations",
+#     (1, 10, 127),
+# )
+def test_mixtral_model_inference(
```
Comment (mtairum) [inline models/demos/t3000/mixtral8x7b/tests/test_mixtral_model.py:49]: Uncomment the pytest parametrizations.

File: models/demos/t3000/mixtral8x7b/tests/test_mixtral_rms_norm.py [hunk 2]
```diff
@@ -23,32 +24,39 @@
...
+        dtype=dtype,
+        layout=ttnn.TILE_LAYOUT,
+        mesh_mapper=ReplicateTensorToMesh(t3k_device_mesh),
+    )
+    tt_input = ttnn.to_device(tt_input, t3k_device_mesh)
     tt_output = tt_model(tt_input)
-    tt_output_torch = ttnn.to_torch(tt_output).squeeze(0)
-
+    tt_output_torch = ttnn.to_torch(tt_output, mesh_composer=ConcatMeshToTensor(t3k_device_mesh, dim=0))[0]
+    print(tt_output_torch.shape, reference_output.shape)
```
Comment (mtairum) [inline models/demos/t3000/mixtral8x7b/tests/test_mixtral_rms_norm.py:59]: Remove print

File: models/demos/t3000/mixtral8x7b/tests/test_mixtral_rms_norm.py [hunk 3]
```diff
@@ -23,32 +24,39 @@
...
     dtype = ttnn.bfloat8_b
-    model_args = TtModelArgs(device)
-    state_dict = torch.load(model_args.consolidated_weights_path(0))
+
+    model_args = TtModelArgs(t3k_device_mesh.get_device(0))
+    state_dict = torch.load(model_args.state_dict_path)
 
     # Ref model needs partial state dict, but our models use full state dict keys as cached weight names
-    partial_state_dict = {k[24:]: v for k, v in state_dict.items() if (k.startswith("layers.0.attention_norm."))}
+    partial_state_dict = {k[24:]: v for k, v in state_dict.items() if (k.startswith("layers.7.attention_norm."))}
```
Comment (mtairum) [inline models/demos/t3000/mixtral8x7b/tests/test_mixtral_rms_norm.py:34]: Is this intented or a typo?

**Review by mtairum — CHANGES_REQUESTED** [review]
Please fix this issues before we merge.

**Comment by mtairum** [comment]
New pipelines for the fixes running here:

multi-device unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/9109035704 
multi-device demo: https://github.com/tenstorrent/tt-metal/actions/runs/9109001480
multi-device perf: https://github.com/tenstorrent/tt-metal/actions/runs/9109039877

**Review by yieldthought — APPROVED** [review]

**Comment by mtairum** [comment]
Running updated pipelines:

https://github.com/tenstorrent/tt-metal/actions/runs/9130971955
https://github.com/tenstorrent/tt-metal/actions/runs/9130974546
https://github.com/tenstorrent/tt-metal/actions/runs/9130975564
https://github.com/tenstorrent/tt-metal/actions/runs/9130976749

After they're all green we should confirm the time it takes to run all `test_mixtral_model.py` tests and adjust them to the correct pipelines and/or reduce the number of tests if needed.

**Review by pgkeller — COMMENTED** [review]
Says I'm tagged as code owner, but I don't see any files on the list that I know anything about (looks like all .py?). Let me know if that's blocking at some point and I'll approve

**Comment by tapspatel** [comment]
Added a comment. Rerun pipelines to ensure passing and ill approve.

**Review by cfjchu — APPROVED** [review]
Looks great!

**Comment by mtairum** [comment]
All pipelines passing here:

- [T3K] T3000 model perf tests: https://github.com/tenstorrent/tt-metal/actions/runs/9161206926
- [T3K] T3000 frequent tests: https://github.com/tenstorrent/tt-metal/actions/runs/9161205037
- [T3K] T3000 unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/9161210132
- [T3K] T3000 demo tests: https://github.com/tenstorrent/tt-metal/actions/runs/9172485831

I'll squash some commits and merge.

**Review by mtairum — APPROVED** [review]

**Review by tapspatel — APPROVED** [review]

**Pull request merged**